### PR TITLE
Handes visibility preference setting for dm/gm

### PR DESCRIPTION
**Pull request by shazm**
#### Summary

The mobile app now handles DM/GM visibility preferences (when a user has selected to "close" the conversation).

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-42641

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on:

- Android (SIM) Pixel 5a API29

#### Release Note

```release-note

```

### CODE REVIEW

File: app/components/channel_list/categories/body/index.ts [hunk 1]
```diff
@@ -82,19 +85,37 @@ const getSortedIds = (database: Database, category: CategoryModel, locale: strin
     }
 };
 
-const enhance = withObservables(['category'], ({category, locale, database}: {category: CategoryModel; locale: string} & WithDatabaseArgs) => {
+const mapPrefName = (prefs: PreferenceModel[]) => of$(prefs.map((p) => p.name));
+
+const mapChannelIds = (channels: ChannelModel[]) => of$(channels.map((c) => c.id));
+
+const enhance = withObservables(['category'], ({category, locale, database, currentUserId}: {category: CategoryModel; locale: string; currentUserId: string} & WithDatabaseArgs) => {
```
Comment (larkox) [inline app/components/channel_list/categories/body/index.ts:92]: Nit: The type wouldn't look better in a "OwnProps" type?

File: app/components/channel_list/categories/categories.tsx [hunk 2]
```diff
@@ -24,15 +25,17 @@ const styles = StyleSheet.create({
...
     const intl = useIntl();
+
     const renderCategory = useCallback((data: {item: CategoryModel}) => {
         return (
             <>
                 <CategoryHeader category={data.item}/>
                 <CategoryBody
                     category={data.item}
                     currentChannelId={currentChannelId}
+                    currentUserId={currentUserId}
```
Comment (larkox) [inline app/components/channel_list/categories/categories.tsx:38]: If category body has already an index file that has to go either way to the database, why not get it from the database there, instead of drilling it down here?

File: app/constants/preferences.ts [hunk 3]
```diff
@@ -11,6 +11,7 @@ const Preferences: Record<string, any> = {
     CATEGORY_FAVORITE_CHANNEL: 'favorite_channel',
     CATEGORY_AUTO_RESET_MANUAL_STATUS: 'auto_reset_manual_status',
     CATEGORY_NOTIFICATIONS: 'notifications',
+
```
Comment (larkox) [inline app/constants/preferences.ts:14]: Unrelated change?

File: app/queries/servers/preference.ts [hunk 4]
```diff
@@ -65,3 +67,17 @@ export const deletePreferences = async (database: ServerDatabase, preferences: P
         return false;
     }
 };
+
+export const queryPreferencesByCategory = (database: Database, category: string, name?: string, value?: string) => {
+    const clauses = [Q.where('category', category)];
+
+    if (typeof name === 'string') {
```
Comment (larkox) [inline app/queries/servers/preference.ts:74]: I like better the `!= null` option to make sure it is not undefined, but probably just personal preference.

**Review by larkox — APPROVED** [review]
LGTM. Just a few nits.

File: app/components/channel_list/categories/categories.tsx (hunk 2 above)
Comment (enahum) [reply app/components/channel_list/categories/categories.tsx:38]: The idea was to prevent having to perform the same query (regardless of being cached) for each category when the parent can perform the query just once

File: app/queries/servers/preference.ts (hunk 4 above)
Comment (shazm) [reply app/queries/servers/preference.ts:74]: `typeof` covers all three cases; skipping `name = null` and `name = undefined` but still passing true for `name = ''`.

**Review by enahum — APPROVED** [review]

File: app/queries/servers/preference.ts (hunk 4 above)
Comment (larkox) [reply app/queries/servers/preference.ts:74]: `!= null` also cover the three cases: https://stackoverflow.com/questions/38977829/what-is-the-difference-between-null-and-null
(see that it is just one `=` and not two.
